shared/dataflow/codeql/dataflow/internal/ContentDataFlowImpl.qll

```diff
@@ -455,27 +455,42 @@ module MakeImplContentDataFlow<LocationSig Location, InputSig<Location> Lang> {
       )
     }
 
+    pragma[nomagic]
+    private predicate nodeAndState(Flow::PathNode n, Node node, State state) {
+      n.getNode() = node and n.getState() = state
+    }
+
+    pragma[nomagic]
+    private predicate succNodeAndState(
+      Flow::PathNode pre, Node preNode, State preState, Flow::PathNode succ, Node succNode,
+      State succState
+    ) {
+      nodeAndState(pre, preNode, preState) and
+      nodeAndState(succ, succNode, succState) and
+      pre.getASuccessor() = succ
+    }
```

Comment:
I don't think you need both of those, right? I think if inline one of them (either should work) then you'd still prevent the bad join. I haven't checked completely, but that's what it looks like to me at a quick glance.

Reply:
I will just try and inline `nodeAndState` in the other predicate.
It is not possible to inline the content of `succNodeAndState` in the predicates where it is used.